Document LDAP support (#19099)
cockroachdb/docs: src/current/v24.3/ldap-authentication.md

---
title: Configure LDAP Authentication

> summary: Learn how to configure CockroachDB to use LDAP for user authentication with directory services like Active Directory.

Reviewer: ```suggestion
summary: Learn how to authenticate to a cluster using identities stored in LDAP and or in a directory service such as Azure Active Directory.
```

toc: true
---

> CockroachDB supports authentication and authorization using LDAP (Lightweight Directory Access Protocol) with services such as Active Directory. This allows you to integrate your existing directory service with CockroachDB for centralized user management and access control.

Reviewer: ```suggestion
CockroachDB supports authentication and authorization using identities stored in LDAP (Lightweight Directory Access Protocol) with services such as Azure Active Directory. This allows you to integrate your cluster with your organization's existing identity infrastructure, allowing for centralized user management and access control.
```

Author: ```suggestion
CockroachDB supports authentication and authorization using LDAP-compatible directory services, such as Active Directory and Microsoft Entra ID. This allows you to integrate your cluster with your organization's existing identify infrastructure, enabling centralized user management and access control.
```

Further edit for consideration. I figure we should integrate the [new name for Azure AD](https://learn.microsoft.com/en-us/entra/fundamentals/new-name), while also making reference to the more well-known Active Directory (which still exists / is still true; per that link "Active Directory" remains correct terminology as a shorthand for "Windows Server Active Directory".

Reviewer: Cool, I didn't know there was a new name. My instinct to add "Azure" was for SEO and that I'd be very surprised if our ICP tend to run internal non-Azure Active Directory instances that they connect to their clusters. I guess I could be wrong.

> This page describes how to configure CockroachDB user authentication using LDAP. You can also configure user authorization (role-based access control) with LDAP, to set users' CockroachDB roles based on group memberships in the directory, but this is not required.

Reviewer: I got lost in the end of this para, but I think maybe it means that you can authenticate to your cluster using identities stored in LDAP, and you can assign them roles based on their group membership.  I wonder if you could assign them roles based on other criteria, such as their existence in the directory.

> This page describes how to configure CockroachDB user authentication using LDAP. You can also configure user authorization (role-based access control) with LDAP, to set users' CockroachDB roles based on group memberships in the directory, but this is not required.

Author: ```suggestion
This page describes how to configure CockroachDB user authentication using LDAP. You can additionally configure CockroachDB to use the same directory service for user authorization (role-based access control), assigning CockroachDB roles to users based on their group memberships in the directory.
```

Reviewer: Maybe "and assign" to avoid a gerund. This is more clear, thank you.

## Overview

> LDAP authentication in CockroachDB uses a 'search and bind' approach:

Reviewer: ```suggestion
LDAP authentication in CockroachDB uses a "search and bind" approach:
```
I've never heard this specific terminology before. Worth linking to a wiki or similar?

> 1. CockroachDB first connects to the LDAP server using a service account (bind DN).

Reviewer: ```suggestion
1. The cluster connects to the LDAP server using a service account (bind DN).
```
I'm confused by the parenthetical. Are we saying that the LDAP source must have its bind DN configured in some specific way to point to a cloud infrastructure service account?

Reviewer 2: This is correct, but is there a need to get into more specifics? Actually we have 1 connection per CockroachDB cluster node, so if there are n nodes in the cluster, we consume n ldap connections on LDAP AD service. Also to note these connections are established only if there is a request to a particular node for connecting via LDAP and are not closed unless AD server closes it.

Reviewer: I'm trying to get really clear about it because the bind DN effectively sets the search scope for the LDAP search, right? Thanks for the extra details.

Reviewer 2: > bind DN effectively sets the search scope for the LDAP search

Yes, in 1 authN attempt, we first bind as bindDN, perform search for user based on HBA attr and `sAMAccountName` provided as sql username and get user DN.
We then use the userDN over the same connection to bind to AD to see if creds were correct.
Post that we again rebind over same connection as bindDN and search for the groups the user is assigned to.
All of this is for a single auth attempt to single CRDB cluster node.

Reviewer: This makes sense to me, thank you. There are several actors here so I think it's important to talk about the cluster, the directory service, the authenticating user very specifically. It would be easy to get lost in who is doing what.

> 2. It searches for the authenticating user using configurable search criteria.

Reviewer: ```suggestion
1. The cluster searches within the directory for the a record that matches the authenticating user, using configurable search criteria.
```

Author: Please refer to my suggestion on the first step.

> 3. If found, it attempts to bind to LDAP using the user's credentials.

Reviewer: ```suggestion
1. If a record is found, the cluster attempts to bind to LDAP using the credentials provided by the user.
```

Author: Please refer to my suggestion on the first step.

> 4. Upon successful binding, the user is authenticated to CockroachDB.

Reviewer: ```suggestion
4. If binding is successful, this authenticates the user to CockroachDB.
```

Author: Please refer to my suggestion on the first step.

> This method supports both standard LDAP servers and Active Directory, with secure LDAPS connections over TLS.

Reviewer: ```suggestion
LDAP authentication be used with both standard LDAP servers and services such as Azure Active Directory. Secure LDAPS connectivity over TLS is required.
```
But maybe it belongs in Prerequisites, which is nearby already.

Author: ```suggestion
LDAP authentication in CockroachDB works with any LDAP-compatible directory service, including Microsoft Active Directory and OpenLDAP. Secure LDAPS connectivity over TLS is supported, and recommended for production environments.
```

> LDAP authentication in CockroachDB works with any LDAP-compatible directory service, including Microsoft Entra ID, Active Directory, and OpenLDAP. Secure LDAPS connectivity over TLS is supported, and recommended for production environments.

Author: @souravcrl While I think this is what you're saying, I am still not 100% clear: is it true that every connection/request *must* use LDAPS?

> LDAP authentication in CockroachDB works with any LDAP-compatible directory service, including Microsoft Entra ID, Active Directory, and OpenLDAP. Secure LDAPS connectivity over TLS is supported, and recommended for production environments.

Reviewer 2: Yes that's right. The above comment is correct
```
Secure LDAPS connectivity over TLS is supported, and recommended for production environments.
```

{{site.data.alerts.callout_info}}

> LDAP authentication cannot be used for the `root` user or other admin users. These credentials must be managed separately using password authentication to ensure continuous administrative access, especially if the LDAP service becomes unavailable.

Reviewer: ```suggestion
LDAP authentication is not supported for the `root` user or other admin users. These credentials must be managed separately using password authentication to ensure continuous administrative access if the LDAP server or service becomes unavailable.
```
What do we mean by "admin users"? Users with the admin role, or an identity literally called "admin"?

Reviewer 2: This is the same as the rule for GSSAPI provided here: https://www.cockroachlabs.com/docs/v21.2/gssapi_authentication#configuring-the-cockroachdb-node
refer to pt.7
```
Setting the server.host_based_authentication.configuration [cluster setting](https://www.cockroachlabs.com/docs/v21.2/cluster-settings) to this particular value makes it mandatory for all non-root users to authenticate using GSSAPI. The root user is always an exception and remains able to authenticate using a valid client cert or a user password.
```
The rule is enforced by unstated unchangeable root access rule mentioned here: https://www.cockroachlabs.com/docs/stable/security-reference/authentication#the-unstated-unchangeable-root-access-rule

> LDAP authentication is not supported for the `root` user or other admin users.
This needs a slight change. We cannot use root user or other reserved roles for LDAP authN. 
Reserved roles can be a node user or have prefix `pg_` or `crdb_internal_`.  Node users are not supposed to be authenticating via cluster SSO/db console. `pg_` and `crdb_internal_` are special roles I think which are used for running cluster jobs or some specific operations are not created directly(need to verify this).

Author: @souravcrl [Unless you object] I'm just going to say that LDAP cannot be used to authenticate the root user or other users with the admin role. I don't think we need to mention `node`, `pg_*` and `crdb_internal*` unless questions arise.

Side note: is there a similar setting here as in GSSAPI that *only* allows users (with the above root/admin exceptions) to auth via LDAP; no fallback? (If I'm reading GSSAPI correctly.) Or is that something we might add later?

Reviewer 2: > I'm just going to say that LDAP cannot be used to authenticate the root user or other users with the admin role.

We can just say
LDAP cannot be used to authenticate the root user or other reserved identities.

> is there a similar setting here as in GSSAPI that only allows users (with the above root/admin exceptions) to auth via LDAP; no fallback?

There is no such setting currently for GSSAPI. Only the HBA conf mandates a rule for root user to have cert-password auth which supersedes all other HBA entries. For admin user there is no such requirement. Hence GSSAPI only enforces root user cannot be used. Admin users can still use GSSAPI.

Admin users can use LDAP also. But we don't need to mention it as it is not enforced/mandated anywhere.

{{site.data.alerts.end}}

## Configuration

### Prerequisites

> 1. An LDAP or Active Directory server accessible from your CockroachDB nodes.

Reviewer: ```suggestion
1. An LDAP or Active Directory server accessible from your CockroachDB nodes.
```
I think "accessible from your CockroachDB nodes" can probably be removed.

> 2. A service account (bind DN) with permissions to search the directory.

Reviewer: I'm still not clear on the service account -- is it an AD service account that you are configuring CockroachDB to use to bind to the LDAP server, or is it a different kind of service account? I haven't seen "service account" used for self-hosted installations very often, we tend to use it for CC on GCP.

Nit: Prerequisites should probably be bullets, npot a numbered list.

3. Network connectivity on port 636 for LDAPS.

> 4. The LDAP server's CA certificate if using LDAPS.

Reviewer: Does this mean you need to download the certificate or get it into your cluster somehow? 

What port will we use if it is not LDAPS?

> ### Step 1: Set sensitive cluster settings to be redacted

Reviewer: ```suggestion
### Step 1: Enable redaction of sensitive cluster settings
```

> LDAP configuration requires you to store LDAP bind credentials and the mapping of external identities to SQL users within CockroachDB cluster settings (`server.host_based_authentication.configuration`, `server.identity_map.configuration`). Prior to this configuration, ensure that your cluster is set to redact sensitive cluster settings for all users except those who are assigned the `admin` role or were granted the `MODIFYCLUSTERSETTING` privilege.

Reviewer: ```suggestion
You will configure the LDAP bind credentials and the mapping of external identities to CockroachDB SQL users using the [cluster settings]({% link {{ page.version.version }}/cluster-settings.md %}) `server.host_based_authentication.configuration` and `server.identity_map.configuration`.

To redact these sensitive setting values for all users except for those with the `admin` role or the `MODIFYCLUSTERSETTING` privilege, run:
```

This is missing the copy block at the beginning of the code block.

~~~ sql
SET CLUSTER SETTING server.redact_sensitive_settings.enabled = 'true';
~~~

> ### Step 2: Configure Host-Based Authentication (HBA)

Reviewer: Do we have a good place to link to for more details about HBA? What if this section is just called "Enable LDAP host-based authentication" and then you explode HBA in the first para? This will make the anchor more SEO-friendly.

Reviewer 3: Probably worth linking to the HBA page
https://www.cockroachlabs.com/docs/stable/security-reference/authentication#hba-configuration-syntax

Reviewer 3: Probably worth linking to the HBA page
https://www.cockroachlabs.com/docs/stable/security-reference/authentication#hba-configuration-syntax

> Set the `server.host_based_authentication.configuration` cluster setting to enable LDAP authentication.

Reviewer: ```suggestion
The configuration for LDAP host-based authentication (HBA) is set in the `server.host_based_authentication.configuration` cluster setting.
```

> The format for HBA is `TYPE  DATABASE  USER  ADDRESS  METHOD  [OPTIONS]`. For LDAP, the method should be set to `ldap` and the option parameters that follow will be LDAP specific:

Reviewer: I'm having a hard time mapping this `TYPE DATABASE....` format to the following list. Are we saying that `type` must be set to `host`, `database` must be set to `all`, `user` must be set to `all`, etc? I don't know what to do with the format for HBA, I guess. I see the link to the security reference, but I am not sure there are enough details here.

Reviewer 3: Maybe we should avoid re-explaning here and just link the HBA docs
https://www.cockroachlabs.com/docs/stable/security-reference/authentication#hba-configuration-syntax

- `ldapserver`: LDAP server hostname
- `ldapport`: LDAP server port (typically 636 for LDAPS)
- `ldapbasedn`: Base DN for user searches
- `ldapbinddn`: Service account DN for directory searches
- `ldapbindpasswd`: Service account password
- `ldapsearchattribute`: Attribute to match against SQL usernames
- `ldapsearchfilter`: LDAP filter to restrict valid users

For example:

~~~ sql
SET CLUSTER SETTING server.host_based_authentication.configuration = '
host    all    all    all    ldap    ldapserver=ldap.example.com 
    ldapport=636 
    "ldapbasedn=ou=users,dc=example,dc=com" 
    "ldapbinddn=cn=readonly,dc=example,dc=com" 
    ldapbindpasswd=readonly_password 
    ldapsearchattribute=uid 
    "ldapsearchfilter=(memberof=cn=cockroachdb_users,ou=groups,dc=example,dc=com)"';
~~~

If you also intend to [configure LDAP Authorization]({% link v24.3/ldap-authorization.md %}#configuration), you will need to include an additional LDAP parameter, `ldapgrouplistfilter`.

> For more information on the HBA fields and HBA in CockroachDB, generally, refer to [SQL Authentication configuration](https://www.cockroachlabs.com/docs/stable/security-reference/authentication#authentication-configuration).

Reviewer: This needs a Liquid link.
```suggestion
For more information about CockroachDB's support for HBA and for more information about the HBA fields and format,, refer to [SQL Authentication configuration](https://www.cockroachlabs.com/docs/stable/security-reference/authentication#authentication-configuration).
```

### Step 3: Configure TLS (Optional)

If using LDAPS with a certificate signed by a custom Certificate Authority (CA) that is not in the system's trusted CA store, you'll need to configure the CA certificate. This step is only necessary when using certificates signed by your organization's private CA or other untrusted CA.

Set the custom CA certificate:

> ~~~ sql

Reviewer: Needs copy block

SET CLUSTER SETTING server.ldap_authentication.domain_ca = '<PEM_ENCODED_CA_CERT>';
~~~

Configure a client certificate for mTLS if required:

> ~~~ sql

Reviewer: Needs copy block

SET CLUSTER SETTING server.ldap_authentication.client.tls_certificate = '<PEM_ENCODED_CERT>';
SET CLUSTER SETTING server.ldap_authentication.client.tls_key = '<PEM_ENCODED_KEY>';
~~~

### Step 4: Sync Database Users

> Before LDAP authentication can be used for a user, the username must be created directly in CockroachDB. You will, therefore, need to establish an automated method for keeping users in sync with the directory server, creating and dropping them as needed.

Reviewer: ```suggestion
Before LDAP authentication can be used for a user, the username must be created directly in CockroachDB. You must establish an automated method for keeping users in sync with the directory server, including fcreating and dropping users from the cluster as needed.
```

> For Azure AD deployments, the username typically corresponds to the `sAMAccountName` field from AD's person object, which is specified in the HBA configuration using `ldapsearchattribute=sAMAccountName`.

Reviewer: Do we want to give any other examples besides Azure AD? Want to use the new name for AAD?

Reviewer 3: We should call it Microsoft Entra ID (formerly called Azure AD)
https://www.microsoft.com/en-us/security/business/identity-access/microsoft-entra-id

Author: I looked further into this, and it looks like:

- For Self-Managed Active Directory: 
  - It's the `user` object (not `person` object), `sAMAccountName` field. These are short and typically compatible with our username requirements.
  
- For Entra ID (formerly Azure AD):
  - The `user` object instead has a `UserPricinpalName` field, considered to be the replacement for `sAMAccountName`.
  - This can't be used as-is because it has an `@` symbol (`<local_name>@<domain>`).
  - As an alternative, a customer can likely use `mailNickname`. Here's how it is set for a user:
    - When created through Azure Portal: `mailNickname` is automatically generated based on the user's entered username (the local part of what will become their UPN).
    - When synced from self-managed AD: The `mailNickname` is typically derived from the on-premises attributes. It often matches the `sAMAccountName` from self-managed AD.
    - When created programmatically (Graph API, PowerShell):`mailNickname` is a required attribute, but no automatic matching happens; the customer is responsible for setting it.
  - Or, I suppose the customer could set up a new field just for specifying a CRDB username by splitting or otherwise transforming the UPN.
    
I haven't fully verified all of these details so I'm happy to leave it at the Active Directory `sAMAccountName` example, for now, with some slight refinements, unless @souravcrl or @dikshant would prefer to confirm/include an Entra ID recommendation here.

Author: Thank you. Despite what I said on Slack, if you think it's safe to assume it's there for all/most Entre ID instances, we might as well put it back in. Or we can wait if more verification would be helpful.

{{site.data.alerts.callout_info}}

> SQL usernames must comply with CockroachDB's [username requirements](https://www.cockroachlabs.com/docs/v24.2/create-user#user-names). Ensure that the values in the LDAP `sAMAccountName` field (or other mapped attribute) meet these requirements.

Reviewer: Liquid link

{{site.data.alerts.end}}

To create a single user:

> ~~~ sql

Reviewer: Needs copy block

CREATE ROLE username LOGIN;
~~~

To create users in bulk:

1. Export usernames from the directory server.
1. Produce a `.sql` file with a [`CREATE ROLE`](https://www.cockroachlabs.com/docs/v24.2/create-role) statement per user, each on a separate line.

>     ~~~ sql

Reviewer: Maybe needs copy block?

(Same feedback for remaining code blocks that you want users to copy and adapt)

    CREATE ROLE username1 LOGIN;
    CREATE ROLE username2 LOGIN;
    CREATE ROLE username3 LOGIN;
    ~~~

    If you are not also enabling LDAP Authorization to manage roles and privileges, you can also include one or more `GRANT` lines for each user. For example, `GRANT developer TO username1` or `GRANT SELECT ON DATABASE orders TO username2;`.

1. Run the SQL file to create the users.

    ~~~ sql
    cockroach sql --file=create_users.sql --host=<servername> --port=<port> --user=<user> --database=<db> --certs-dir=path/to/certs
    ~~~

> You can similarly [`DROP`](https://www.cockroachlabs.com/docs/stable/drop-role) users as they are removed from the directory, and update roles and groups as needed, if not using LDAP authorization.

Reviewer: I'd instead say "To drop a user:" and then give a command, and then maybe something like "You can drop users in bulk in the same way as adding them in bulk.

I guess you could drop and add users in the same SQL file.

> ### Connecting to CockroachDB

Reviewer: ```suggestion
### Connect to a cluster using LDAP
```
I'd make this an H2

> #### SQL shell connection with LDAP authentication

Reviewer: ```suggestion
#### SQL shell
```
I'd make this an H3

> After LDAP authentication is configured, users can connect using their LDAP credentials:

Reviewer: ```suggestion
To connect using LDAP credentials:
```

> ~~~ bash

Reviewer: ```suggestion
~~~ shell
```

# Method 1: Password in connection string
cockroach sql --url "postgresql://username:ldap_password@host:26257" --certs-dir=certs

# Method 2: Password in environment variable
export PGPASSWORD='ldap_password'
cockroach sql --url "postgresql://username@host:26257" --certs-dir=certs

> ~~~

Reviewer: I would not bury this detail in comments. I'd pull it out into para text:

You can store the LDAP password directly in the connection string, or you can set it in the `PGPASSWORD` environment variable before connecting.

Are there security concerns about either approach here? The password will be in your shell history if you use either of these options. It's a more valuable password than just a cluster one, because it potentially gets you into the Azure tenant itself.

Reviewer 3: Passing passwords/secrets through env vars is a pretty common practice.

#### DB Console connection with LDAP authentication

If LDAP authentication is configured, DB Console access will also use this configuration, allowing users to log in with their SQL username and LDAP password. During a login attempt, the system checks if LDAP authentication is configured for the user in the HBA configuration. If so, it validates the credentials against the LDAP server. If LDAP authentication fails or is not configured, the system falls back to password authentication.

> {{site.data.alerts.callout_info}}

Reviewer: I'm not sure I understand this. I think that granting access to DB Console is separate from granting it to a SQL client, right?

Reviewer 3: We have made some changes. Previously the DB console was using a different auth flow. Now when using LDAP, if LDAP is configured for the SQL user they can also login to the DB Console using that same set of credentials.
